File: backend/app/routers/alerts.py
from fastapi import APIRouter, Depends, Request
from fastapi.responses import HTMLResponse, RedirectResponse
from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.db import get_db
from app.deps import current_user
from app.models import Alert, ApiKey, User
from app.templating import templates

router = APIRouter(tags=["alerts"])


@router.get("/alerts", response_class=HTMLResponse)
async def alerts_page(
    request: Request,
    user: User = Depends(current_user),
    db: AsyncSession = Depends(get_db),
):
    rows = (
        await db.execute(
            select(Alert, ApiKey.alias)
            .outerjoin(ApiKey, Alert.api_key_id == ApiKey.id)
            .where(Alert.user_id == user.id)
            .order_by(Alert.created_at.desc())
            .limit(200)
        )
    ).all()
    return templates.TemplateResponse(
        request, "alerts.html", {"user": user, "alerts": rows}
    )


@router.post("/alerts/read")
async def mark_all_read(
    user: User = Depends(current_user), db: AsyncSession = Depends(get_db)
):
    await db.execute(update(Alert).where(Alert.user_id == user.id).values(read=True))
    await db.commit()
    return RedirectResponse("/alerts", status_code=303)

File: src/routes/alerts.js
const express = require("express");
const { Alert, ApiKey } = require("../models");
const { requireAuth } = require("../middleware/auth");
const { asyncHandler } = require("../middleware/asyncHandler");

const router = express.Router();

router.get(
  "/alerts",
  requireAuth,
  asyncHandler(async (req, res) => {
    const rows = await Alert.findAll({
      where: { userId: req.user.id },
      include: [{ model: ApiKey, as: "key", attributes: ["alias"], required: false }],
      order: [["createdAt", "DESC"]],
      limit: 200,
    });

    // Template expects a list of [alert, alias] pairs, matching the
    // Python router's `select(Alert, ApiKey.alias).outerjoin(...)` rows.
    const alerts = rows.map((alert) => [alert, alert.key ? alert.key.alias : null]);

    res.render("alerts.html", { user: req.user, alerts });
  })
);

router.post(
  "/alerts/read",
  requireAuth,
  asyncHandler(async (req, res) => {
    await Alert.update({ read: true }, { where: { userId: req.user.id } });
    res.redirect(303, "/alerts");
  })
);

module.exports = router;
